src/arch/riscv/pmp.cc

```diff
@@ -0,0 +1,191 @@
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "arch/riscv/pmp.hh"
+#include "arch/generic/tlb.hh"
+#include "base/addr_range.hh"
+#include "base/types.hh"
+#include "math.h"
+#include "mem/request.hh"
+#include "params/PMP.hh"
+#include "sim/sim_object.hh"
+
+PMP::PMP(const Params &params) :
+SimObject(params), maxEntries(params.max_pmp)
+{
+
+    for (int i=0; i < maxEntries; i++) {
+        pmpEntry* entry = new pmpEntry;
+        AddrRange thisRange(0,-1);
+        entry->pmpAddr = thisRange;
+        entry->pmpCfg = 0;
+        pmpTable.push_back(entry);
+    }
+
+}
+
+// check if this region can be
+// accessed (else raise an exception)
+bool
+PMP::pmpCheck(const RequestPtr &req, BaseTLB::Mode mode)
+{
+
+    std::cout << "pmp: check" << std::endl;
+
+    // all pmp entries need to be looked from the lowest to
+    // the highest number
+
+    bool addrMatch = false;
+
+    int i;
+    for (i = 0; i < pmpTable.size(); i++) {
+        std::cout << "pmp: table check " << i << "   " << std::endl;
+        if (pmpTable[i]->pmpAddr.contains(req->getPaddr())){
+           //address matched
+            addrMatch = true;
+            break;
+        }
+    }
+
+    if (!addrMatch)
+        return false; //no pmp entry has a matching address
+
+
+    // else check the RWX permissions from the pmp entry
+
+    uint8_t allowedPrivs = PMP_READ | PMP_WRITE | PMP_EXEC;
+
+    // i is the index of pmp table which matched
+    allowedPrivs &= pmpTable[i]->pmpCfg;
+
+    std::cout << "pmp: priv check " << std::endl;
+
+    if ((mode == BaseTLB::Mode::Read) &&
+            ((PMP_READ & allowedPrivs) == PMP_READ))
+        return true;
+    else if ((mode == BaseTLB::Mode::Write) &&
+            ((PMP_WRITE & allowedPrivs) == PMP_WRITE))
+        return true;
+    else if ((mode == BaseTLB::Mode::Execute) &&
+            ((PMP_WRITE & allowedPrivs) == PMP_EXEC))
+        return true;
```

Comment:
```suggestion
    if ((mode == BaseTLB::Mode::Read) && (PMP_READ & allowedPrivs))
        return true;
    else if ((mode == BaseTLB::Mode::Write) && (PMP_WRITE & allowedPrivs))
        return true;
    else if ((mode == BaseTLB::Mode::Execute) && (PMP_WRITE & allowedPrivs))
        return true;
```
90% sure that's right :). The `allowed_privs & MASK` should be true iif it matches.

Reply:
Yeah, this is correct (because the mask will only have 1 bit set at a time). fixed!